ethcore/src/engines/authority_round/mod.rs

```diff
@@ -815,10 +816,11 @@ impl AuthorityRound {
 					None
 				}
 			})
-				.while_some();
+				.while_some()
+				.take_while(|&(h, _)| h != epoch_transition_hash);
```

Comment:
So that limits the iterator, can you elaborate a little bit why it's valid to do so?

Reply:
I discussed this with @ngotchac. Aura only used to track finality for safely triggering validator set changes, therefore we were only interested in finalizing blocks within the current epoch, so this `take_while` was the original behavior. Recently we added finality tracking to the client so we updated Aura to track finality until the last finalized block. On chains where there isn't a majority of validators online the unfinalized subchain just keeps growing and this can take some time to build (although if the subchain is properly cached this cost is avoided).